### PR TITLE
ENH: Adding new pandas option for integer formatting, Issue#57177

### LINKED ISSUE #57177: ENH: Expose integer formatting also for default display, not only styler
### Feature Type

- [X] Adding new functionality to pandas

- [ ] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

I noticed that in Pandas version `2.2` the internal implementation of  `IntArrayFormatter` was changed (or at least renamed), marking the classes as private using one underscore. I used the implementation as described [here](https://stackoverflow.com/a/29663750/9684872) to format integers in a DataFrame on each display.

See changes between `2.1` and `2.2` to make the linked patch work: [link](https://github.com/RasmussenLab/njab/pull/4/commits/10b0c84ddb28a1dd9c2bd58e9b6af148dbd6c9dc)

I see that the styler supports formatting thousands:

```python
import pandas as pd
pd.options.styler.format.thousands = ','
s = pd.Series([1_000_000]).to_frame()
s.style # does change display of s to 1,000,000
```
however
```
s # will display 1000000
```
Would it be possible to add the same to the display options?

### Feature Description

pd.options.format.thousands = ','

would need to pass:

```python
if pd.options.format.thousands:
  format_str = f':{pd.options.format.thousands}d'.format
else:
  format_str = ':d'.format
```

### Alternative Solutions

Change the default [here](https://github.com/pandas-dev/pandas/blob/main/pandas/io/formats/format.py#L1475)

```python
class IntArrayFormatter(pf._GenericArrayFormatter):

    def _format_strings(self):
        formatter = self.formatter or '{:,d}'.format
        fmt_values = [formatter(x) for x in self.values]
        return fmt_values

pf._IntArrayFormatter = IntArrayFormatter
```

Using the above patch I get the display I would prefer:

```
>>> pd.Series([1_000_000])
0    1,000,000
dtype: int64
```

### Additional Context

Working with large integers in tables is made easier if thousands are visually distinguished. Maybe there is already a way to specify that formatter function (`str.format`) for display, which would be perfect.

### PR DESCRIPTION
**Pull request by Meadiocre**
- [ ] closes #57177  (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Modified the _IntArrayFormatter class and added a new display option to format integers with a delimiter in DataFrames. The default value for the delimiter is '' and user have the option of using ',' or '_' for integer separator. Note this change does not affect float or complex numbers.

### CODE REVIEW

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -95,6 +95,7 @@
         CompressionOptions,
         FilePath,
         FloatFormatType,
+        IntegerFormatType,
```
Comment (Dr-Irv) [inline pandas/io/formats/format.py:98]: I'm surprised this worked because `_typing.py` has no `IntegerFormatType`

Comment (Meadiocre) [reply pandas/io/formats/format.py]: Oh did not notice that until I created a PR, just pushed something that fixes that.

**Comment by Meadiocre** [comment]
@Dr-Irv Any idea why I am getting this precommit error? I ran "pre-commit run --hook-stage manual --all-files" locally with no errors but I am getting one when I push my changes. 

image

**Comment by Dr-Irv** [comment]
> @Dr-Irv Any idea why I am getting this precommit error? I ran "pre-commit run --hook-stage manual --all-files" locally with no errors but I am getting one when I push my changes.
> 
> image

If you look at the output of the CI check here (https://results.pre-commit.ci/run/github/858127/1714714926.f5LU8a9SRWKOAlR2F43xpQ), that will indicate the issue.  I'm not sure if the pre-commit checks run locally are the same as run in CI.

**Comment by Meadiocre** [comment]
@Dr-Irv Forgot to mention that I looked at that which lead me to post the comment. Just could not figure out what this error meant. Tried to see if this ruff-format ran locally so that I could commit the file that was reformatted, but it did not happen. 

image

**Comment by Dr-Irv** [comment]
> @Dr-Irv Forgot to mention that I looked at that which lead me to post the comment. Just could not figure out what this error meant. Tried to see if this ruff-format ran locally so that I could commit the file that was reformatted, but it did not happen.
> 
> image

Make sure you do the following:
1) Merge with `upstream/main`
2) Update your environment, e.g., `mamba env update --file environment.yml `
3) Update the pre-commit hooks:  `pre-commit install`

I did this, edited a file that I knew would get fixed by `ruff-format`, and it was invoked when committing that file.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**